evap/rewards/forms.py

```diff
@@ -1,14 +1,108 @@
+from contextlib import contextmanager
+from datetime import date
+
 from django import forms
+from django.core.exceptions import ValidationError
+from django.core.validators import MaxValueValidator, StepValueValidator
+from django.db import transaction
+from django.utils.translation import gettext as _
 
-from evap.rewards.models import RewardPointRedemptionEvent
+from evap.evaluation.models import UserProfile
+from evap.rewards.models import RewardPointRedemption, RewardPointRedemptionEvent
+from evap.rewards.tools import reward_points_of_user
 
 
 class RewardPointRedemptionEventForm(forms.ModelForm):
     class Meta:
         model = RewardPointRedemptionEvent
-        fields = ("name", "date", "redeem_end_date")
+        fields = ("name", "date", "redeem_end_date", "step")
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.fields["date"].localize = True
         self.fields["redeem_end_date"].localize = True
+
+
+class RewardPointRedemptionForm(forms.Form):
+    event = forms.ModelChoiceField(queryset=RewardPointRedemptionEvent.objects.all(), widget=forms.HiddenInput())
+    points = forms.IntegerField(min_value=0, label="")
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if not self.initial:
+            return
+
+        self.fields["points"].validators.append(MaxValueValidator(self.initial["total_points_available"]))
+        self.fields["points"].widget.attrs["max"] = self.initial["total_points_available"]
+
+        self.fields["points"].validators.append(StepValueValidator(self.initial["event"].step))
+        self.fields["points"].widget.attrs["step"] = self.initial["event"].step
+
+        if self.initial["event"].step > 1:
+            self.fields["points"].help_text = _("multiples of {}").format(self.initial["event"].step)
+
+    def clean_event(self):
+        event = self.cleaned_data["event"]
+        if event.redeem_end_date < date.today():
+            raise ValidationError(_("Sorry, the deadline for this event expired already."))
+        return event
+
+
+class BaseRewardPointRedemptionFormSet(forms.BaseFormSet):
+    def __init__(self, *args, user: UserProfile, **kwargs) -> None:
+        self.user = user
+        super().__init__(*args, **kwargs)
+        self.locked = False
+
+    def get_form_kwargs(self, index):
+        kwargs = super().get_form_kwargs(index)
+        if not self.initial:
+            return kwargs
+        kwargs["initial"] = self.initial[index]
+        kwargs["initial"]["total_points_available"] = reward_points_of_user(self.user)
```

Comment:
Doesn't need to be inside the critical section. This only ends up in the `MaxValueValidator` in `RewardPointRedemptionForm`, which is only there for frontend/usability. For (security) correctness, only the `total_points_available = reward_points_of_user(self.user)` inside `BaseRewardPointRedemptionFormSet.clean` is relevant.

Reply:
But just to be sure, these validators also run later when handling the POST data right? Just with maybe outdated bounds

Reply:
It's just a django form field validator -- yes, it better be validated when calling `is_valid`.

The reward points being outdated could theoretically be an issue if a user was awarded points in the meantime. The `MaxValueValidator` allows providing a callable instead of a value. Maybe that allows shifting the computation to a time inside the locked phase -- but for just rendering the page, we never lock any rows anyway...